compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/types/BallerinaFunctionTypeBuilder.java

```diff
@@ -0,0 +1,209 @@
+package io.ballerina.compiler.api.impl.types;
+
+import io.ballerina.compiler.api.impl.SymbolFactory;
+import io.ballerina.compiler.api.impl.symbols.AbstractTypeSymbol;
+import io.ballerina.compiler.api.impl.symbols.BallerinaSymbol;
+import io.ballerina.compiler.api.impl.symbols.TypesFactory;
+import io.ballerina.compiler.api.symbols.FunctionTypeSymbol;
+import io.ballerina.compiler.api.symbols.ParameterKind;
+import io.ballerina.compiler.api.symbols.ParameterSymbol;
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.SymbolTable;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BInvokableTypeSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BVarSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.SymTag;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.Symbols;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BInvokableType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BType;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+import org.wso2.ballerinalang.compiler.util.Names;
+import org.wso2.ballerinalang.util.Flags;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.ballerinalang.model.symbols.SymbolOrigin.COMPILED_SOURCE;
+
+/**
+ * The implementation of the methods used to build the Function type descriptor in Types API.
+ *
+ * @since 2.0.0
+ */
+public class BallerinaFunctionTypeBuilder implements TypeBuilder.FUNCTION {
+
+    private final CompilerContext context;
+    private final TypesFactory typesFactory;
+    private final SymbolTable symTable;
+    private List<ParameterSymbol> parameterSymbols = new ArrayList<>();
+    private ParameterSymbol restParam;
+    private TypeSymbol returnTypeSymbol;
+
+    public PARAMETER_BUILDER PARAM_BUILDER;
+
+    public BallerinaFunctionTypeBuilder(CompilerContext context) {
+        this.context = context;
+        typesFactory = TypesFactory.getInstance(context);
+        symTable = SymbolTable.getInstance(context);
+        PARAM_BUILDER = new ParameterBuilder(context);
+    }
+
+    @Override
+    public TypeBuilder.FUNCTION withParams(ParameterSymbol... parameters) {
+        if (!parameterSymbols.isEmpty()) {
+            parameterSymbols = new ArrayList<>();
+        }
```

Comment:
What is the reason for this?

Reply:
Similar to https://github.com/ballerina-platform/ballerina-lang/pull/36422#discussion_r908233441
Improved this using the `.clear()` method.